pkg/odo/cli/debug/portforward.go

```diff
@@ -0,0 +1,149 @@
+
+	"github.com/openshift/odo/pkg/util"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+
+	corev1 "k8s.io/api/core/v1"
+
+	"k8s.io/kubernetes/pkg/kubectl/cmd/templates"
+	k8sgenclioptions "k8s.io/kubernetes/pkg/kubectl/genericclioptions"
+)
+
+// PortForwardOptions contains all the options for running the port-forward cli command.
+type PortForwardOptions struct {
+	Namespace string
+	// PortPair is the combination of local and remote port in the format "local:remote"
+	PortPair string
+
+	localPort  int
+	contextDir string
+
+	PortForwarder *debug.DefaultPortForwarder
+	// StopChannel is used to stop port forwarding
+	StopChannel chan struct{}
+	// ReadChannel is used to receive status of port forwarding ( ready or not ready )
+	ReadyChannel chan struct{}
+	*genericclioptions.Context
+	localConfigInfo *config.LocalConfigInfo
+}
+
+var (
+	portforwardLong = templates.LongDesc(`
+			Forward a local port to a remote port on the pod where the application is listening for a debugger.
+
+			By default the local port and the remote port will be same. To change the local port use can use --local-port argument and to change the remote port use "odo config set DebugPort <port>"   		  
+	`)
+
+	portforwardExample = templates.Examples(`
+		# Listen on default port and forwarding to the default port in the pod
+		odo debug port-forward 
+
+		# Listen on the 5000 port locally, forwarding to default port in the pod
+		odo debug port-forward --local-port 5000
+		
+		`)
+)
+
+const (
+	portforwardCommandName = "port-forward"
+)
+
+func NewPortForwardOptions() *PortForwardOptions {
+	return &PortForwardOptions{}
+}
+
+// Complete completes all the required options for port-forward cmd.
+func (o *PortForwardOptions) Complete(name string, cmd *cobra.Command, args []string) (err error) {
+
+	o.Context = genericclioptions.NewContext(cmd)
+	cfg, err := config.NewLocalConfigInfo(o.contextDir)
+	o.localConfigInfo = cfg
+
+	remotePort := cfg.GetDebugPort()
+	o.PortPair = fmt.Sprintf("%d:%d", o.localPort, remotePort)
+
+	// Using Discard streams because nothing important is logged
+	o.PortForwarder = debug.NewDefaultPortForwarder(o.Context.Client, k8sgenclioptions.NewTestIOStreamsDiscard())
+
+	o.StopChannel = make(chan struct{}, 1)
+	o.ReadyChannel = make(chan struct{})
+	return nil
+}
+
+// Validate validates all the required options for port-forward cmd.
+func (o PortForwardOptions) Validate() error {
+
+	if len(o.PortPair) < 1 {
+		return fmt.Errorf("ports cannot be empty")
+	}
+	return nil
+}
+
+// Run implements all the necessary functionality for port-forward cmd.
+func (o PortForwardOptions) Run() error {
+	componentName := o.localConfigInfo.GetName()
+	appName := o.localConfigInfo.GetApplication()
+	componentLabels := componentlabels.GetLabels(componentName, appName, false)
+	componentSelector := util.ConvertLabelsToSelector(componentLabels)
+	dc, err := o.Client.GetOneDeploymentConfigFromSelector(componentSelector)
```

Comment:
These client calls should be in the ForwardPorts() function and the client should be sent as a parameter to the function.

Reply:
Actually no because forwardports should have no information about if its running on kubernetes or openshift

Reply:
According to me, we have always avoided using occlient directly in the cli package, so I guess we can handle the abstraction in some other form in the debug package itself. This seems to break the current architecture.

Reply:
Also in future this might create confusion as it deviates from the current pattern

Reply:
done @mik-dass

Reply:
now all the occlient related constructs are part of the portforward package